subsys/bluetooth/host/mesh/models/generic_battery_srv.c

```diff
@@ -0,0 +1,89 @@
+/* Bluetooth Mesh Models */
+
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <net/buf.h>
+
+#include <bluetooth/mesh.h>
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_MESH_DEBUG_MODEL)
+#include "common/log.h"
+
+static inline void u32_to_le24(u32_t value, u8_t *u24)
+{
+	u24[0] = value & 0xff;
+	u24[1] = (value >> 8) & 0xff;
+	u24[2] = (value >> 16) & 0xff;
+}
+
+/* Generic Battery Server message handlers */
+static void gen_battery_get(struct bt_mesh_model *model,
+			    struct bt_mesh_msg_ctx *ctx,
+			    struct net_buf_simple *buf)
+{
+	struct net_buf_simple *msg = NET_BUF_SIMPLE(2 + 8 + 4);
+	struct generic_battery_state *state = model->user_data;
+	u8_t u24_buff[3];
+
+	bt_mesh_model_msg_init(msg, BT_MESH_MODEL_OP_2(0x82, 0x24));
+	net_buf_simple_add_u8(msg, state->battery_level);
+	u32_to_le24(state->time_to_discharge, u24_buff);
+	net_buf_simple_add_mem(msg, u24_buff, sizeof(u24_buff));
+	u32_to_le24(state->time_to_charge, u24_buff);
+	net_buf_simple_add_mem(msg, u24_buff, sizeof(u24_buff));
+	net_buf_simple_add_u8(msg, state->flags);
+
+	if (bt_mesh_model_send(model, ctx, msg, NULL, NULL)) {
+		BT_ERR("Unable to send Battery Status response");
+	}
+}
+
+/* Mapping of message handlers for Generic battery messages (0x100c) */
+const struct bt_mesh_model_op bt_mesh_model_gen_battery_srv_op[] = {
+	{ BT_MESH_MODEL_OP_2(0x82, 0x23), 0, gen_battery_get},
+	BT_MESH_MODEL_OP_END,
+};
+
+struct generic_battery_state
+	generic_battery_state_user_data[CONFIG_BT_MESH_MODEL_BATTERY_SRV_CNT];
+
+NET_BUF_SIMPLE_DEFINE_STATIC(bt_mesh_pub_msg_generic_battery_srv, 10);
+
+struct bt_mesh_model_pub
+	generic_battery_pub[CONFIG_BT_MESH_MODEL_BATTERY_SRV_CNT] = { [0 ...
+	(CONFIG_BT_MESH_MODEL_BATTERY_SRV_CNT - 1)] = {
+		.msg = &bt_mesh_pub_msg_generic_battery_srv }
```

Comment:
This can't really work if you have multiple batteries sending out different messages at the same time. Each battery should have its own unique message.

Reply:
Is it possible to publish at same time from many batteries ?

Reply:
I suppose so. Also be aware of the Publish Retransmit Count state, since that may mean that the model's message will keep being used for some time after the bt_mesh_model_publish() call returns.

Reply:
Ok, sounds reasonable to do separate publication msg buffers.